File: providers/tristar/src/main/kotlin/gdynia/provider.kt
package app.wheretopark.providers.tristar.gdynia

import app.wheretopark.providers.shared.Provider
import app.wheretopark.shared.*
import com.charleskorn.kaml.Yaml
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlin.time.Duration
import kotlin.time.Duration.Companion.days
import kotlin.time.Duration.Companion.seconds

const val METADATA_URL = "http://api.zdiz.gdynia.pl/ri/rest/parkings"
const val STATE_URL = "http://api.zdiz.gdynia.pl/ri/rest/parking_places"

@Serializable
data class ParkingLotConfiguration(
    val resources: List<ParkingLotResource>,
    @SerialName("total-spots")
    val totalSpots: UInt,
    val features: List<ParkingLotFeature>,
    val rules: List<ParkingLotRule>
)

@Serializable
data class Configuration(
    @SerialName("parking-lots")
    val parkingLots: Map<UInt, ParkingLotConfiguration>,
)

class TristarGdyniaProvider : Provider() {
    override val name: String
        get() = "tristar/gdynia"
    override val metadataInterval: Duration
        get() = 1.days
    override val stateInterval: Duration
        get() = 30.seconds

    private val client = HttpClient {
        install(ContentNegotiation) {
            json(Json {
                prettyPrint = true
                isLenient = true
                ignoreUnknownKeys = true
            })
        }
    }

    // Mapping of vendor and storekeeper ID's
    private val ids = mutableMapOf<UInt, ParkingLotID>()

    private val configuration =
        Yaml().decodeFromString<Configuration>(this.javaClass.getResource("/configuration-gdynia.yaml")!!.readText())

    override suspend fun metadatas(): Map<ParkingLotID, ParkingLotMetadata> {
        val vendorMetadatas = client.get(METADATA_URL).body<MetadataResponse>()
        return vendorMetadatas.parkingLots.map {
            val configuration = configuration.parkingLots[it.id] ?: return@map null
            val id = it.location.coordinates.hash()
            ids[it.id] = id
            val metadata = ParkingLotMetadata(
                name = it.name,
                address = it.address,
                location = it.location.coordinates,
                resources = configuration.resources,
                totalSpots = mapOf(
                    Pair(ParkingSpotType.CAR, configuration.totalSpots)
                ),
                features = configuration.features,
                currency = "PLN",
                rules = configuration.rules,
            )
            id to metadata
        }
            .filterNotNull()
            .toMap()
    }

    override suspend fun states(): Map<ParkingLotID, ParkingLotState> {
        val vendorStates = client.get(STATE_URL).body<StateResponse>()
        return vendorStates.map {
            val id = ids[it.parkingID] ?: return@map null
            val state = ParkingLotState(
                lastUpdated = it.insertTime,
                availableSpots = mapOf(
                    Pair(ParkingSpotType.CAR, it.freePlaces)
                ),
            )
            id to state
        }
            .filterNotNull()
            .toMap()
    }
}

File: providers/tristar/gdynia/provider.go
package gdynia

import (
	"fmt"
	"log"
	"time"
	wheretopark "wheretopark/go"
)

type Provider struct {
	configuration Configuration
	mapping       map[int]wheretopark.ID
}

func (p Provider) GetMetadata() (map[wheretopark.ID]wheretopark.Metadata, error) {
	vendorMetadata, err := GetMetadata()
	if err != nil {
		return nil, err
	}
	metadatas := make(map[wheretopark.ID]wheretopark.Metadata)
	for _, vendor := range vendorMetadata.Parkings {
		configuration, exists := p.configuration.ParkingLots[vendor.ID]
		if !exists {
			log.Printf("missing configuration for %d\n", vendor.ID)
			continue
		}
		id := wheretopark.GeometryToID(vendor.Location)
		metadata := wheretopark.Metadata{
			Name:           vendor.Name,
			Address:        vendor.Address,
			Geometry:       vendor.Location,
			Resources:      configuration.Resources,
			TotalSpots:     configuration.TotalSpots,
			MaxWidth:       configuration.MaxWidth,
			MaxHeight:      configuration.MaxHeight,
			Features:       configuration.Features,
			PaymentMethods: configuration.PaymentMethods,
			Comment:        configuration.Comment,
			Currency:       "PLN",
			Timezone:       "Europe/Warsaw",
			Rules:          configuration.Rules,
		}
		metadatas[id] = metadata
		p.mapping[vendor.ID] = id
	}
	return metadatas, nil
}

func (p Provider) GetState() (map[wheretopark.ID]wheretopark.State, error) {
	vendorState, err := GetState()
	if err != nil {
		return nil, err
	}
	states := make(map[wheretopark.ID]wheretopark.State)
	for _, vendor := range *vendorState {
		id, exists := p.mapping[vendor.ParkingID]
		if !exists {
			log.Printf("no mapping for %d\n", vendor.ID)
			continue
		}

		fmt.Printf("vendorstate: %+v\n", vendor)

		lastUpdate, err := time.Parse("2006-01-02 15:04:05", vendor.InsertTime)
		if err != nil {
			return nil, err
		}
		state := wheretopark.State{
			LastUpdated: lastUpdate.Format(time.RFC3339),
			AvailableSpots: map[string]uint{
				"CAR": uint(vendor.FreePlaces),
			},
		}
		states[id] = state
	}
	return states, nil
}

func NewProvider(configurationPath *string) (*Provider, error) {
	var configuration Configuration
	if configurationPath == nil {
		configuration = DefaultConfiguration
	} else {
		newConfiguration, err := LoadConfiguration(*configurationPath)
		if err != nil {
			return nil, err
		}
		configuration = *newConfiguration
	}
	return &Provider{
		configuration: configuration,
		mapping:       make(map[int]wheretopark.ID),
	}, nil

}
